strawberry/federation/schema.py

```diff
@@ -354,33 +322,54 @@ def _warn_for_federation_directives(self) -> None:
         pass
 
 
-def _get_entity_type(type_map: "TypeMap") -> Optional[GraphQLUnionType]:
-    # https://www.apollographql.com/docs/apollo-server/federation/federation-spec/#resolve-requests-for-entities
+def _get_entity_type(
+    query: Optional[Type[WithStrawberryObjectDefinition]],
+    mutation: Optional[Type[WithStrawberryObjectDefinition]],
+    subscription: Optional[Type[WithStrawberryObjectDefinition]],
+    additional_types: Iterable[Type[WithStrawberryObjectDefinition]],
+) -> Optional[StrawberryUnion]:
+    # recursively iterate over the schema to find all types annotated with @key
+    # if no types are annotated with @key, then the _Entity union and Query._entities
+    # field should not be added to the schema
 
-    # To implement the _Entity union, each type annotated with @key
-    # should be added to the _Entity union.
+    entity_types = set()
 
-    federation_key_types = [
-        type.implementation
-        for type in type_map.values()
-        if _has_federation_keys(type.definition)
-        # TODO: check this
-        and not isinstance(type.implementation, GraphQLInterfaceType)
-    ]
+    # need a stack to keep track of the types we need to visit
+    stack: List[Any] = [query, mutation, subscription, *additional_types]
 
-    # If no types are annotated with the key directive, then the _Entity
-    # union and Query._entities field should be removed from the schema.
-    if not federation_key_types:
-        return None
+    seen = set()
+
+    while stack:
+        type_ = stack.pop()
+
+        if type_ is None:
+            continue
+
+        while isinstance(type_, StrawberryContainer):
+            type_ = type_.of_type
 
-    entity_type = GraphQLUnionType("_Entity", federation_key_types)  # type: ignore
+        if not hasattr(type_, "__strawberry_definition__"):
+            continue
 
-    def _resolve_type(self: Any, value: Any, info: Any) -> str:
-        return self.__strawberry_definition__.name
+        type_definition = get_object_definition(type_, strict=True)
+
+        if type_definition.is_object_type and _has_federation_keys(type_definition):
+            entity_types.add(type_)
+
+        for field in type_definition.fields:
+            if field.type and field.type not in seen:
+                seen.add(field.type)
+                stack.append(field.type)
```

Comment:
nit: reduce one indentation here (but ignore if you prefer the other way)

```suggestion
            if field.type and field.type in seen:
                continue

            seen.add(field.type)
            stack.append(field.type)
```